### PR TITLE
Schema registry article

### PR DESCRIPTION
**Pull request by cleahy3**
article discussing our implementation of the schema registry

### CODE REVIEW

File: _posts/2017-03-17-implementing-the-schema-registry.md [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+tags:       Data, Improvements
+author:     callum_leahy
+---
+
+![Enigma Machine](/images/enigma.jpg)
+
+## Introduction
+
+The platform currently used within the Gaming Promotions Squad largely relies on Apache Kafka. Kafka is a data streaming service which is used to send data from one section of the stack to another. Apache Kafka works with Apache Avro, a data serialization system used to encode Kafka. As a result of using Avro we need to use a Schema Registry Service, this service will store the various Avro schemas that we require in order to encode and decode messages.
+
```
Comment (ls12styler) [inline _posts/2017-03-17-implementing-the-schema-registry.md]: In the paragraph above, I'd change "system used to encode Kafka" to be "system used to encode event data placed on to Kafka

Comment (ls12styler) [inline _posts/2017-03-17-implementing-the-schema-registry.md]: Also maybe link to the relevant project pages too.

File: _posts/2017-03-17-implementing-the-schema-registry.md [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+## What it is
+
+The Confluent Schema Registry is an idea by the creators of the Kafka platform. It is a secure and safe way of storing schema versions and ensuring that an accurate history is kept in case a rollback is needed. The schema registry is stored in the _schema topic, which is on Kafka itself. So anyone who has access to Kafka will therefore have access to the schema registry. This gives us the GPS squad end-to-end schema availability meaning we can massively increase the reliability of a system.
```
Comment (ls12styler) [inline _posts/2017-03-17-implementing-the-schema-registry.md]: "This gives us the GPS squad end-to-end schema"
Should be (note the commas, I think it reads better that way):
"This gives us, the GPS squad, end-to-end schema"

Also, not sure about the use of 'GPS Squad' - that essentially says 'Gaming Promotions Squad Squad'.

File: _posts/2017-03-17-implementing-the-schema-registry.md [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+## How we do it
+
+Our Registry client was built to spec with the [Confluent Schema Registry]( 
+http://docs.confluent.io/3.1.1/schema-registry/docs/). For the development of the client we have used NodeJS; using a Node package ensures it is reusable to deploy by other teams. When looking into the use of Node we realized that there were disadvantages, such as language native functions, specifically data serialization. However, we believed this was offset by the obvious advantage of having the use of NPM. NPM would be used to combat the native deserialization functionality by using the [AVSC package](https://github.com/mtth/avsc). This package is an implementation of the Avro which specifically uses JavaScript. This use of JavaScript allows commonality, thus making it easier to encode and decode messages sent in Avro, provided the correct schema is given.
```
Comment (ls12styler) [inline _posts/2017-03-17-implementing-the-schema-registry.md]: Towards the end you have "implementation of the Avro which", maybe this should read "implementation of the Avro specification, which"

File: _posts/2017-03-17-implementing-the-schema-registry.md [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+The main part to the Schema Registry implementation is the client, it is a layer on top of the Schema Registry Restful Service. The Node package interacts with the Registry Service to return schemas to the client. Once retrieved the package then caches the schemas to ensure its as efficient as possible when it comes to timings and network resources. Once cached it then passes the schema to another package, which in turn is resolved into a promise of an encoded or decoded message back to the consumer of the package.
```
Comment (ls12styler) [inline _posts/2017-03-17-implementing-the-schema-registry.md]: "ensure its as efficient", the "its" should be "it's" (it is)

File: _posts/2017-03-17-implementing-the-schema-registry.md [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+The Confluent Schema Registry has a specific format in which messages should be encoded, this is called the 'Wire Format'. The wire format is essentially a way of ensuring that all encoded messages sent include some form of identification in how they were encoded for the schema to read. The format contains one 'Magic Byte', known as the Confluent serialization format version number, this is currently always 0. Additionally, this is followed by the schema Id which has a format of 4 Bytes.
```
Comment (ls12styler) [inline _posts/2017-03-17-implementing-the-schema-registry.md]: Link to the 'Wire Format" documentation?

Comment (ls12styler) [inline _posts/2017-03-17-implementing-the-schema-registry.md]: The casing on "Id" in the last sentence. Maybe just uppercase both letters? "ID"?

Comment (ls12styler) [inline _posts/2017-03-17-implementing-the-schema-registry.md]: I'm not sure the sentence "The wire format is essentially a way of ensuring that all encoded messages sent include some form of identification in how they were encoded for the schema to read." reads that well. Suggestions?

File: _posts/2017-03-17-implementing-the-schema-registry.md [hunk 6]
```diff
@@ -0,0 +1,40 @@
+---
+layout:     post
+title:      Implementing The Schema Registry
+date:       2017-02-08 12:00
```
Comment (WheresAlice) [inline _posts/2017-03-17-implementing-the-schema-registry.md]: the date in the filename does not match this date

File: _posts/2017-03-17-implementing-the-schema-registry.md [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+summary:    Keeping on top of an area of technology that is as rapidly moving as the big data ecosystem is hard.  Our data tribe share some of their resources for keeping up to date.
```
Comment (WheresAlice) [inline _posts/2017-03-17-implementing-the-schema-registry.md]: inconsistent spacing (2 spaces after the first sentence)

File: _posts/2017-03-17-implementing-the-schema-registry.md [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+
+The Confluent Schema Registry has a specific format in which messages should be encoded, this is called the 'Wire Format'. The wire format is essentially a way of ensuring that all encoded messages sent include some form of identification in how they were encoded for the schema to read. The format contains one 'Magic Byte', known as the Confluent serialization format version number, this is currently always 0. Additionally, this is followed by the schema Id which has a format of 4 Bytes.
+
+## Overview
+
+The Schema Registry is essential to teams using Apache Avro and Kafka to stream data from one end to another, the registry makes the encoding and decoding of these messages much more reliable and easy to collaborate on. The registry holds all different versions of a schema and can be accessed by anyone relevant to the kafka stream. The schemas are then used to encode and decode avro encoded messages, usually from Kafka.
```
Comment (ls12styler) [inline _posts/2017-03-17-implementing-the-schema-registry.md]: Just ensure the casing of 'avro' is consistent throughout the post.

**Comment by WheresAlice** [comment]
I think I'd like to see a bit more detail, but that could be a followup post.  For example, what happens when I publish a message that doesn't follow the schema?  A bit about making forward compatible changes to the schema would be good too.

**Comment by ls12styler** [comment]
@WheresAlice I think Schema Evolution is definitely worth its own post. It's a pretty large topic in itself (the Registry does actually enforce forward/backwards compatibility - dependant on the subjects config).

File: _posts/2017-03-17-implementing-the-schema-registry.md [hunk 9]
```diff
@@ -0,0 +1,41 @@
+---
+layout:     post
+title:      Implementing The Schema Registry
+date:       2017-03-17 12:00
+summary:    Keeping on top of an area of technology that is as rapidly moving as the big data ecosystem is hard. Our data tribe share some of their resources for keeping up to date.
```
Comment (WheresAlice) [inline _posts/2017-03-17-implementing-the-schema-registry.md]: that summary sounds awfully familiar

Comment (ls12styler) [inline _posts/2017-03-17-implementing-the-schema-registry.md]: LOL

File: _posts/2017-03-17-implementing-the-schema-registry.md [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+
+
+## Introduction
+
+The platform currently used within the Gaming Promotions Squad largely relies on Apache Kafka. Kafka is a data streaming service which is used to send data from one section of the stack to another. [Apache Kafka](https://kafka.apache.org/intro) works with [Apache Avro](http://avro.apache.org/docs/current/), a data serialization system used to encode event data placed onto Kafka. As a result of using Avro we need to use a Schema Registry Service, this service will store the various Avro schemas that we require in order to encode and decode messages.
+
+## What it is
+
+The Confluent Schema Registry is an idea by the creators of the Kafka platform. It is a secure and safe way of storing schema versions and ensuring that an accurate history is kept in case a rollback is needed. The schema registry is stored in the _schema topic, which is on Kafka itself. So anyone who has access to Kafka will therefore have access to the schema registry. This gives us the Gaming Promotions squad, end-to-end schema availability meaning we can massively increase the reliability of a system.
```
Comment (ls12styler) [inline _posts/2017-03-17-implementing-the-schema-registry.md]: Casing on "Gaming Promotions squad" should be "Squad"

File: _posts/2017-03-17-implementing-the-schema-registry.md [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+
+## How we do it
+
+Our Registry client was built to spec with the [Confluent Schema Registry]( 
+http://docs.confluent.io/3.1.1/schema-registry/docs/). For the development of the client we have used NodeJS; using a Node package ensures it is reusable to deploy by other teams. When looking into the use of Node we realized that there were disadvantages, such as language native functions, specifically data serialization. However, we believed this was offset by the obvious advantage of having the use of NPM. NPM would be used to combat the native deserialization functionality by using the [AVSC package](https://github.com/mtth/avsc). This package is an implementation of the Avro, which specifically uses JavaScript. This use of JavaScript allows commonality, thus making it easier to encode and decode messages sent in Avro, provided the correct schema is given.
```
Comment (ls12styler) [inline _posts/2017-03-17-implementing-the-schema-registry.md]: "implementation of the Avro, which" - Can this be changed to read "implementation of the Avro specification, which".

Comment (ls12styler) [inline _posts/2017-03-17-implementing-the-schema-registry.md]: Or maybe even just get rid of "the".

**Comment by ls12styler** [comment]
@cleahy3 If you amend that sentence and the post date to today, I'll publish it for you.

**Pull request merged**